Schedules tasks

vorayash opened the pull request:
Pull Request for Issue Schedules Tasks #37313 

### Summary of Changes
Changed "joomla-cms\administrator\components\com_scheduler\forms\task.xml" by making cron fields required so that if user tries to save task without checking all the field he/she will get warning.


### Testing Instructions

1. Create Demo Task - Sleep
2. Select Cron Expression Advanced as Execution Rule
3. Do not select all values in area Crone match.


### Actual result BEFORE applying this Pull Request
![image](https://user-images.githubusercontent.com/72155000/159555477-7bfb7ae6-b83f-4a17-ac79-e974d83286c9.png)



### Expected result AFTER applying this Pull Request
![image](https://user-images.githubusercontent.com/72155000/159555825-fab409a0-0bc1-4c76-b43a-3634208a2229.png)

vorayash committed:
added required in cron-expression fields

vorayash committed:
removed spaces

richard67 commented:
What happens if you chose another method than cron expression in the drop down? The cron expression fields should not be required in this case.

vorayash committed:
added condition for cron-fields

it will check if execution-rule is "cron-expression" then there should be all the field available in cron-expression array otherwise give warning.

vorayash committed:
removed required from cron fields

vorayash committed:
Merge branch '4.1-dev' into Schedules_Tasks

vorayash commented:
I have removed required from cron expression and have modified FormField.php. Please check.

richard67 commented:
@vorayash Your PR has code style errors. You can see that when going to the bottom of the PR and see the drone CI is failing. With the "Details" link you can get to the details.

![2022-03-23_pr-37350_drone_1](https://user-images.githubusercontent.com/7413183/159690606-480a6b0f-ea88-4799-9a57-700a69c80e97.png)

When following to that link you come to a page where on the left hand side you can see the failed PHPCS test. By selecting that test, the right hand side shows the log from the test, which tells the details about the code style errors:

![2022-03-23_pr-37350_drone_2](https://user-images.githubusercontent.com/7413183/159690578-30106439-d247-4163-b533-2a06b4d65ae0.png)

Could you fix them? If you need help let me know, then I will propose a code change which you can accept with one mouse click. But I wanted to give you the chance first to fix them yourself.

Thanks in advance.

richard67 commented:
P.S.: You should adjust your text editor or IDE editor so that it shows you white space like spaces and tabulators. This helps to avoid the white space at the end of line errors.

vorayash commented:
I will surely do it. Thank you sir for your support.

vorayash committed:
removed whitespace

vorayash committed:
removed whitespace

vorayash committed:
Removed whitespace

vorayash committed:
Removed whitespace

vorayash commented:
Please check this.

Quy commented on libraries/src/Form/FormField.php:
```diff
@@ -1287,6 +1287,21 @@ public function validate($value, $group = null, Registry $input = null)
+		$execution_rules = (array) $input->get('execution_rules');
+
+		// Check if execution rule is cron-expression
+		if ($execution_rules['rule-type'] === "cron-expression")
```
```suggestion
		if ($execution_rules['rule-type'] === 'cron-expression')
```

vorayash committed:
Update libraries/src/Form/FormField.php

Co-authored-by: Quy <quy@fluxbb.org>

richard67 commented:
I think the testing instructions are not complete. It does not only need to test if the PR does what is required for the task form where it has a field set "execution_rules", it also needs to be tested that nothing is broken in forms which use the FormField class or a class derived from it and do NOT have that field set.

I think in this case we will get PHP errors related to unset array elements, e.g. `$execution_rules['rule-type']` or `$execution_rules['cron-expression']` and so on.

I haven't tested this PR yet, but I am pretty sure it will raise these PHP errors when doing the additional test which I proposed above.

richard67 commented:
Another question is if it is a good idea to place the checks for this fieldset which only exists in that particular form in a class which is base of many other types of fields used in many other types of forms, so the check is performed in 99.99 % of all cases for nothing.

Maybe it should be placed in the component's model instead, or in the table class for the task table (if we have such, I haven't checked)?

laoneo commented:
Catching the rule error must be added to the scheduler component and not the `FormField` class as it should not be aware about any context. Correctly the controller must pass the valid data to the model as the model and the table should not depend on the input either (I know the model has populatestate where we are suffering since ages from it).

vorayash commented:
Thank you for your guidance. I will solve this issues.

ditsuke commented:
@vorayash are you still working on this?

vorayash commented:
Sir I am right now busy with my GSOC task. Is it possible that after completing it I work on this?

bembelimen commented:
I'm closing this as this is, as stated above, the wrong approach.

If you are available again, please feel free to create a new PR.

vorayash commented:
👍

vorayash commented:
Ok sir.
